collectors/python.d.plugin/mail_log/mail_log.chart.py

```diff
@@ -0,0 +1,234 @@
+            self.queue_ids.pop(msg_id)
+            self.service.stats.accepted += 1
+            self.service.increment_incoming_code("2.0.0")
+
+    def parse_smtpd_and_cleanup(self, line, first, service):
+        """Parses smtpd and cleanup logs"""
+        if service == "smtpd" and first == "connect":
+            self.service.stats.connections += 1
+            return
+        reject = self.REJECT_RE.search(line)
+        if reject:
+            code = reject.group(2)
+            self.service.increment_incoming_code(code)
+            if self.greylist_status and self.greylist_status in line:
+                self.service.stats.greylisted += 1
+            if code.startswith("4"):
+                self.service.stats.temp_rejected += 1
+            elif code.startswith("5"):
+                self.service.stats.perm_rejected += 1
+            return
+        if service == "cleanup":
+            self.queue_ids[first] = None
+            if len(self.queue_ids) > 50:
+                self.queue_ids.popitem(False)
+
+    def parse_smtp(self, line):
+        """Parses smtp client log"""
+        status = self.STATUS_RE.search(line)
+        if status:
+            self.service.increment_outgoing_code(status.group(1))
+            stat = status.group(2)
+            if stat == "sent":
+                self.service.stats.sent += 1
+            elif stat == "deferred":
+                self.service.stats.deferred += 1
+            elif stat == "bounced":
+                self.service.stats.bounced += 1
+
+    def _parse_line(self, line):
+        match = self.POSTFIX_RE.search(line)
+        if not match:
+            return
+        service = match.group(2)
+        first = match.group(3)
+        if service == "qmgr":
+            self.parse_qmgr(line, first)
+        elif service == "smtpd" or service == "cleanup":
+            self.parse_smtpd_and_cleanup(line, first, service)
+        elif service == "smtp":
+            self.parse_smtp(line)
+
+
+class Service(LogService):
+    def __init__(self, configuration=None, name=None):
+        LogService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.stats = MailStatistics()
+        self.parser = {
+            "postfix": PostfixParser
+        }.get(configuration.get("type", "postfix"))(self)
+
+    def check(self):
+        """
+        :return: bool
+
+        1. "log_path" is specified in the module configuration file
+        2. "log_path" must must exist and be readable by netdata user
+        """
+        if not self.log_path:
+            self.error('log path is not specified')
+            return False
+
+        if not (self._find_recent_log_file() and os.access(self.log_path, os.R_OK)):
+            self.error('{log_file} not readable or does not exist'.format(log_file=self.log_path))
+            return False
+        return True
```

Comment:
We need to check if parser is not `None` in `check`. I suggest to move parser init in `check` and log a meaningful error if the type is wrong.

And we need to ensure if the log file actually contains logs of chosen type (postfix, etc.). Not sure how to achieve this. Any ideas?

Reply:
Checking for the presence of matching log lines will give an insanely high false negative rate for whether this should run or not.  Just because someone is running Postfix (or whatever other MTA) does not mean that there will be lines in the mail log from that MTA at any arbitrary point in time (think log rotation).

This is a case that I think we should just rely on the behavior of charts that have never had non-zero values being hidden by default, or just not enable the module by default and let users enable it if they want it.

Reply:
If i understand it right we cant have default parser here. if there is no way to determine MTA we should run the collector, so we cant have it enabled by default. 

> just rely on the behavior of charts that have never had non-zero values being hidden by default

Hm, in this case the module it will silently run forever.

Reply:
The issue I'm worried about is a case where Netdata starts up right after the log files got rotated.  Unless the system deals with lots of mail all the time, it's not unreasonable to expect that there might be no lines whatsoever from the MTA in the log file for quite some time, in which case the module wouldn't start up even though it should be running.

Perhaps disable the module by default, and maybe add a special warning that gets logged if it finds no lines from an MTA in some number of collection cycles (probably 3600 or more)?

Reply:
Oh, i didnt know, so all MTAs write to same file? :see_no_evil: 	

> if it finds no lines from an MTA in some number of collection cycles (probably 3600 or more)?

Well, if the module failes to get data python.d will increase it update every up to 10min gradually. And it will never kill the job (current implementation/strategy - if the job passes check never kill it even if it failed to collect data over 9k times). 

https://github.com/netdata/netdata/blob/adb7026b14fea7bbebee93c8958d92ee9cf1efaf/collectors/python.d.plugin/python_modules/bases/FrameworkServices/SimpleService.py#L52-L55

This is bad because the module will have to read a lot of lines after 10 minutes delay :sweat_smile:

Reply:
My main concern that it will be confusing/wrong to have zero postfix mail log metrics if user has no postfix installed (current implementation).

Reply:
> Oh, i didnt know, so all MTAs write to same file? 🙈

As a general rule, yes, all MTA's (as well as all mail filtering tools and pretty much anything else that doesn't run as a normal user and handles mail) write logs via the syslog `mail` facility, which in standard setups goes to a single file (typically `/var/log/mail.log`).  That's the reason that there's a need to match on the lines being from Postfix itself.

Realistically though, you will only have one MTA running on a system, plus a handful of mail filters.  In general for such a case, you won't git many (if any) log lines when mail is not being processed, so short of this running when the local MTA is something it doesn't support reading log entries from, you shouldn't see much if any slowdown when it fails to collect data.

> My main concern that it will be confusing/wrong to have zero postfix mail log metrics if user has no postfix installed (current implementation).

Agreed, so it sounds like default disabled is probably the best option here.